library(rtracklayer)
library(tidyverse)

gtf_url <- "http://ftp.ensembl.org/pub/release-103/gtf/homo_sapiens/Homo_sapiens.GRCh38.103.chr.gtf.gz"
gtf_path <- ODER:::.file_cache(gtf_url)

# options(stringsAsFactors=F)
# 
# er_file = args[1] # your ers
# gtf_file = args[2]
# expressed_transcript_file = args[3] # file generated by "extract_gtex_expressed_genes.R"
# outpath = args[4]
# tissue <- args[5] # tissue name

gtf <- rtracklayer::import(gtf_path)
gtf <- keepStandardChromosomes(gtf, species = "Homo_sapiens", pruning.mode="coarse")
seqlevelsStyle(gtf) <- "UCSC"   # add chr to seqnames
# gtf.df = as.data.frame(gtf,stringsAsFactor=F)
# gtf.gene = gtf.df %>% dplyr::filter(type %in% "gene")
# gtf.gene.gr <- makeGRangesFromDataFrame(gtf.gene, keep.extra.columns = TRUE)
genesgtf <- gtf[mcols(gtf)[["type"]]=="gene"]
gtf.gene <- as.data.frame(genesgtf)
load("/home/eolagbaju/projects/data/adipose - subcutaneous.Rdata")
adisub_df <- df
load("/home/eolagbaju/projects/data/data.Rdata")
# expressed gtf - extracting only expressed transcripts
#expressed_genes <- read.table(expressed_transcript_file, header=TRUE, sep="\t") #tissue of interest loaded in?
expressed_genes <- data
gtf.exp.gr <- dplyr::semi_join(gtf.gene, expressed_genes, by = c("gene_id" = "Name")) %>%
  makeGRangesFromDataFrame(., keep.extra.columns = TRUE)

er <- read.table(er_file, header=TRUE, sep="\t") %>%
  dplyr::filter(ensembl_grch38_v92_region_annot %in% "intergenic")
###
#aers[S4Vectors::mcols(aers)[["annotation"]] %in% c("intron", "intergenic")]
###
##### add nearest any gene and nearest expressed gene ############
er.gr <- makeGRangesFromDataFrame(er, keep.extra.columns = TRUE)

nearest_hit <- nearest(er.gr, gtf.gene.gr,
                       select=c("arbitrary"), ignore.strand=FALSE)

er$nearest_any_gene_v94_name <- gtf.gene.gr[nearest_hit]$gene_id

exp_hit <- nearest(er.gr, gtf.exp.gr,
                   select=c("arbitrary"), ignore.strand=FALSE)

er$nearest_expressed_gene_v94_name <- gtf.exp.gr[exp_hit]$gene_id
