---
title: "ODER workflow"
author: 
  - name: Emmanuel Olagbaju
    affiliation:
    - UCL
    email: e.olagbaju@ucl.ac.uk
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('ODER')`"
vignette: >
  %\VignetteIndexEntry{Introduction to ODER}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
```


```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library("RefManageR")

# ## Write bibliography information
# bib <- c(
#     R = citation(),
#     BiocStyle = citation("BiocStyle")[1],
#     knitr = citation("knitr")[1],
#     RefManageR = citation("RefManageR")[1],
#     rmarkdown = citation("rmarkdown")[1],
#     sessioninfo = citation("sessioninfo")[1],
#     testthat = citation("testthat")[1],
#     ODER = citation("ODER")[1]
# )
```

# Basics

## Install `ODER`

`R` is an open-source statistical environment which can be easily modified to enhance its functionality via packages. `r Biocpkg("ODER")` is a `R` package available via the [Bioconductor](http://bioconductor.org) repository for packages. `R` can be installed on any operating system from [CRAN](https://cran.r-project.org/) after which you can install `r Biocpkg("ODER")` by using the following commands in your `R` session:

```{r "install", eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("ODER")

## Check that you have a valid Bioconductor installation
BiocManager::valid()
```

## Required knowledge

`r Biocpkg("ODER")` is based on many other packages and in particular in those that have implemented the infrastructure needed for dealing with RNA-seq data (EDIT!). That is, packages like `r Biocpkg("SummarizedExperiment")` (EDIT!).

If you are asking yourself the question "Where do I start using Bioconductor?" you might be interested in [this blog post](http://lcolladotor.github.io/2014/10/16/startbioc/#.VkOKbq6rRuU).


## Asking for help

As package developers, we try to explain clearly how to use our packages and in which order to use the functions. But `R` and `Bioconductor` have a steep learning curve so it is critical to learn where to ask for help. The blog post quoted above mentions some but we would like to highlight the [Bioconductor support site](https://support.bioconductor.org/) as the main resource for getting help: remember to use the `ODER` tag and check [the older posts](https://support.bioconductor.org/t/ODER/). Other alternatives are available such as creating GitHub issues and tweeting. However, please note that if you want to receive help you should adhere to the [posting guidelines](http://www.bioconductor.org/help/support/posting-guide/). It is particularly critical that you provide a small reproducible example and your session information so package developers can track down the source of the error.

## Citing `ODER`

We hope that `r Biocpkg("ODER")` will be useful for your research. Please use the following information to cite the package and the overall approach. Thank you!

```{r "citation"}
## Citation info
# citation("ODER")
```

# Quick start to using to `ODER`

```{r "start", message=FALSE}
library("ODER")
```

```{r loadData}
library(magrittr)
gtex_metadata <- recount::all_metadata("gtex")
# gtex_metadata[["bigwig_file"]][grepl("fibroblast",gtex_metadata[["bigwig_file"]])]
# dplyr::filter(gtex_metadata, bigwig_file %in% "fibroblast")
gtex_metadata <- gtex_metadata %>%
    as.data.frame() %>%
    dplyr::filter(project == "SRP012682")

# obtain path to example bw on recount2
url <- recount::download_study(
    project = "SRP012682",
    type = "samples",
    download = FALSE
)

# grep("fibroblast",url)
# bw_path <- ODER:::.file_cache(url[1])
# bw_path <- ODER:::.file_cache(url[58])
# 84  91   98  106  141  217  245  253  280  301  306  310  318  328  330  375  396  488  522  547  561  588  599
bw_path <- ODER:::.file_cache(url[280])
```

```{r generateERs}

# getting the mean coverage of the expressed regions
coverage <- get_coverage(
    bw_paths = bw_path,
    # auc_raw = gtex_metadata[["auc"]][1],
    # auc_raw = gtex_metadata[["auc"]][58],
    auc_raw = gtex_metadata[["auc"]][280],
    auc_target = 40e6 * 100, # target 40 million coverage with 100 bp length reads
    chrs = c("chr21", "chr22")
) # defaults to chr1-22, chrX, chrY, chrM if chrs = ""

coverage

ers <- get_ers(coverage = coverage, mccs = c(2, 4, 6, 8, 10), mrgs = c(10, 20, 30))
```

```{r non_overlapping_exons}
# download the gtf file from ensembl
gtf_url <- "http://ftp.ensembl.org/pub/release-103/gtf/homo_sapiens/Homo_sapiens.GRCh38.103.chr.gtf.gz"
gtf_path <- ODER:::.file_cache(gtf_url)

exons_no_overlap <- get_exons(gtf = gtf_path, ucsc_chr = TRUE, ignore.strand = TRUE)
```

```{r get_er_deltas}
# get the ers_delta
ers_delta <- get_ers_delta(ers = ers, opt_exons = exons_no_overlap)
# if you have your own delta function you can enter it as an argument i.e.
# delta_fun <- your_delta_function
```

```{r get_opt_ers}
# get the opt_ers
opt_ers <- get_opt_ers(ers = ers, ers_delta = ers_delta)
```

```{r plot_ers, eval=TRUE}
# visualise the spread of mccs and mrgs
opt_plots <- plot_ers(ers_delta = ers_delta, opt_mcc_mrg = opt_ers[["opt_mcc_mrg"]])

opt_plots
```

```{r get_junctions, echo=FALSE}
# load(file = "/data/recount/GTEx_SRP012682/gtex_split_read_table_annotated_rda/lung_split_read_table_annotated.rda")
#
# lung_junc <- gtex_split_read_table_annotated_only_junc_coverage

junctions <- SummarizedExperiment::rowRanges(dasper::junctions_example)
GenomeInfoDb::seqlevelsStyle(junctions) <- "NCBI"

ann_opt_ers <- get_junctions(opt_ers = opt_ers[["opt_ers"]], junc_data = junctions, gtf_path = gtf_path)
genomic_state <- generate_genomic_state(
    gtf = gtf_path,
    chrs_to_keep = c("21", "22"),
    ensembl = TRUE,
)
annot_ers <- derfinder::annotateRegions(
    regions = ann_opt_ers, genomicState = test_gs$fullGenome,
    maxgap = -1L, minoverlap = 1L
)
annot_table <- annotate_table(annot_ers$countTable)
GenomicRanges::mcols(ann_opt_ers)$annotation <- annot_table[["region_annot"]]
GenomicRanges::mcols(ann_opt_ers)$og_index <- annot_table[["ER_index"]]

ann_opt_ers
```

```{r gen_gstate, echo=FALSE}

genomic_state <- generate_genomic_state(
    gtf = gtf_path,
    chrs_to_keep = c("21", "22"),
    ensembl = TRUE,
)
annot_ers <- derfinder::annotateRegions(
    regions = ann_opt_ers, genomicState = test_gs$fullGenome,
    maxgap = -1L, minoverlap = 1L
)
annot_table <- annotate_table(annot_ers$countTable)
GenomicRanges::mcols(ann_opt_ers)$annotation <- annot_table[["region_annot"]]
GenomicRanges::mcols(ann_opt_ers)$og_index <- annot_table[["ER_index"]]

ann_opt_ers
```

```{r annotate_junctions, echo=FALSE}

# load(file = "/data/recount/GTEx_SRP012682/gtex_split_read_table_annotated_rda/lung_split_read_table_annotated.rda")
#
# lung_junc <- gtex_split_read_table_annotated_only_junc_coverage

junctions <- SummarizedExperiment::rowRanges(dasper::junctions_example)
GenomeInfoDb::seqlevelsStyle(junctions) <- "NCBI"
annot_ers <- annotatERs(
    opt_ers = opt_ers[["opt_ers"]], junc_data = junctions,
    gtf_path = gtf_path, chrs_to_keep = c("21", "22"), ensembl = TRUE
)
```

```{r refine_ERs, echo=FALSE}


# # mcols(annot_ers)[["annotation"]][!(grepl("exon",mcols(annot_ers)[["annotation"]]))]
# neon_annot_ers <- annot_ers[mcols(annot_ers)[["annotation"]] %in% c("intron","intergenic")]#remove exon junctions
#
# #neon_annot_ers[!(lengths(mcols(neon_annot_ers)[["grl"]])>2)] #ers where only 1 or 2 junctions overlap
# neon12_annot_ers <- neon_annot_ers[lengths(mcols(neon_annot_ers)[["grl"]])==2 |
#                                      lengths(mcols(neon_annot_ers)[["grl"]])==1]
# # good_js <- lapply(ranges(mcols(neon12_annot_ers)[["grl"]]),inv_colgrs)
# # pro_neon12_annot_ers <- neon12_annot_ers[unlist(good_js)]

# ref_results <- refine_ERs(neon12_annot_ers)

ref_results <- refine_ERs(annot_ers)

refined_ers <- ref_results[[1]]

changes <- ref_results[[2]]

test_grs <- GenomicRanges::GRanges(
    seqnames = S4Vectors::Rle(c("chr21"), c(7)),
    ranges = IRanges::IRanges(
        start = c(5026400, 13482700, 17236500, 18365500, 19350000, 19772900, 20179900),
        end = c(5026500, 13482800, 17609100, 18366000, 19352000, 19773000, 20180000),
        names = c(1:7)
    ),
    strand = S4Vectors::Rle((c("+", "-", "-", "-", "+", "+", "-")))
)

gr1 <- GRanges("chr21", IRanges(5026423, 5323718), strand = "+")
gr2 <- GRanges("chr21", IRanges(c(13261708, 13482787), c(13482707, 13851281)), strand = "-")
gr3 <- GRanges("chr21", IRanges(c(18365503, 18365938), c(18365600, 18857780)), strand = "-")
gr4 <- GRanges("chr21", IRanges(c(18858010, 19351750), c(19350050, 19351960)), strand = "+")
gr5 <- GRanges("chr21", IRanges(19408735, 19772936), strand = "+")
gr6 <- GRanges("chr21", IRanges(20091736, 20179922), strand = "-")
grl <- GenomicRanges::GRangesList(gr1, gr2, gr3, gr4, gr5, gr6)
check_annot_ers <- GenomicRanges::GRanges(
    seqnames = S4Vectors::Rle(c("chr21"), c(6)),
    ranges = IRanges::IRanges(
        start = c(5026400, 13482700, 18365500, 19350000, 19772900, 20179900),
        end = c(5026500, 13482800, 18366000, 19352000, 19773000, 20180000),
        names = c(1:6)
    ),
    strand = S4Vectors::Rle((c("+", "-", "-", "+", "+", "-"))),
    grl = grl
)
```

```{r refine_ERs2, echo=FALSE}

test_grs <- GenomicRanges::GRanges(
    seqnames = S4Vectors::Rle(c("chr21"), c(7)),
    ranges = IRanges::IRanges(
        start = c(5026400, 13482700, 17236500, 18365500, 19350000, 19772900, 20179900),
        end = c(5026500, 13482800, 17609100, 18366000, 19352000, 19773000, 20180000),
        names = c(1:7)
    ),
    strand = S4Vectors::Rle((c("+", "-", "-", "-", "+", "+", "-")))
)

gr1 <- GRanges("chr21", IRanges(5026423, 5323718), strand = "+")
gr2 <- GRanges("chr21", IRanges(c(13261708, 13482787), c(13482707, 13851281)), strand = "-")
gr3 <- GRanges("chr21", IRanges(c(18365503, 18365938), c(18365600, 18857780)), strand = "-")
gr4 <- GRanges("chr21", IRanges(c(18858010, 19351750), c(19350050, 19351960)), strand = "+")
gr5 <- GRanges("chr21", IRanges(19408735, 19772936), strand = "+")
gr6 <- GRanges("chr21", IRanges(20091736, 20179922), strand = "-")
grl <- GenomicRanges::GRangesList(gr1, gr2, gr3, gr4, gr5, gr6)
check_annot_ers <- GenomicRanges::GRanges(
    seqnames = S4Vectors::Rle(c("chr21"), c(6)),
    ranges = IRanges::IRanges(
        start = c(5026400, 13482700, 18365500, 19350000, 19772900, 20179900),
        end = c(5026500, 13482800, 18366000, 19352000, 19773000, 20180000),
        names = c(1:6)
    ),
    strand = S4Vectors::Rle((c("+", "-", "-", "+", "+", "-"))),
    grl = grl
)

test_annot_ers <- annotatERs(
    opt_ers = test_grs, junc_data = junctions,
    gtf_path = gtf_path, chrs_to_keep = c("21", "22"), ensembl = TRUE
)

check_test_annot_ers <- test_annot_ers
changes <- logical(0)
for (a in seq_along(test_annot_ers)) {
    er_start <- start(ranges(test_annot_ers)[a])
    er_end <- end(ranges(test_annot_ers)[a])
    change <- FALSE
    if (length(ranges(mcols(test_annot_ers)[["grl"]])[[a]]) == 1) {
        j_start <- as.integer(start(ranges(mcols(test_annot_ers)[["grl"]])[a]))
        j_end <- as.integer(end(ranges(mcols(test_annot_ers)[["grl"]])[a]))
        if (inbetween(value = j_start, rstart = er_start, rend = er_end) &
            inbetween(value = j_end, rstart = er_start, rend = er_end)) {
            er_start <- j_start + 1
            er_end <- j_end - 1
            change <- TRUE
        } else if (inbetween(value = j_start, rstart = er_start, rend = er_end)) {
            er_end <- j_start - 1
            change <- TRUE
        } else if (inbetween(value = j_end, rstart = er_start, rend = er_end)) {
            er_start <- j_end + 1
            change <- TRUE
        }
    } else if (length(ranges(mcols(test_annot_ers)[["grl"]])[[a]]) == 2) {
        if (colgrs(ranges(mcols(test_annot_ers)[["grl"]])[[a]])) {
            changes <- c(changes, change)
            next
        }
        if (ranges(mcols(test_annot_ers)[["grl"]])[[a]][1] >
            ranges(mcols(test_annot_ers)[["grl"]])[[a]][2]) {
            er_start <- as.integer(end(ranges(mcols(test_annot_ers)[["grl"]])[[a]][2])) + 1
            er_end <- as.integer(start(ranges(mcols(test_annot_ers)[["grl"]])[[a]][1])) - 1
            change <- TRUE
        } else {
            er_start <- as.integer(end(ranges(mcols(test_annot_ers)[["grl"]])[[a]][1])) + 1
            er_end <- as.integer(start(ranges(mcols(test_annot_ers)[["grl"]])[[a]][2])) - 1
            change <- TRUE
        }
    }

    start(check_test_annot_ers[a]) <- er_start
    end(check_test_annot_ers[a]) <- er_end
    changes <- c(changes, change)
}

# mcols(annot_ers)[["annotation"]][!(grepl("exon",mcols(annot_ers)[["annotation"]]))]
ne_annot_ers <- annot_ers[!(grepl("exon", mcols(annot_ers)[["annotation"]]))] # remove exon junctions
neon_annot_ers <- ne_annot_ers[mcols(annot_ers)[["annotation"]] %in% c("intron", "intergenic")] # empty junctions
# neon_annot_ers <- ne_annot_ers[!(mcols(ne_annot_ers)[["annotation"]]=="none")] #empty junctions
# annot_ers[ mcols(annot_ers)[["annotation"]] %in% c("intron","intergenic")]


# neon_annot_ers[!(lengths(mcols(neon_annot_ers)[["grl"]])>2)] #ers where only 1 or 2 junctions overlap
neon12_annot_ers <- neon_annot_ers[lengths(mcols(neon_annot_ers)[["grl"]]) == 2 | lengths(mcols(neon_annot_ers)[["grl"]]) == 1]
good_js <- lapply(ranges(mcols(neon12_annot_ers)[["grl"]]), inv_colgrs)
pro_neon12_annot_ers <- neon12_annot_ers[unlist(good_js)]

# countOverlaps(mcols(neon12_annot_ers)[["grl"]][[2]][1],mcols(neon12_annot_ers)[["grl"]][[2]][2])
# ranges(ne_annot_ers)[1]
# ranges(mcols(ne_annot_ers)[["grl"]])[1]
changes <- logical(0)
# findoverlaps on all junctions, keep er indicies and wrangle the final product into list of TRUEs and FalSEs





for (a in seq_along(neon12_annot_ers)) {
    er_start <- start(ranges(neon12_annot_ers)[a])
    er_end <- end(ranges(neon12_annot_ers)[a])
    change <- FALSE
    if (length(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]]) == 1) {
        j_start <- as.integer(start(ranges(mcols(neon12_annot_ers)[["grl"]])[a]))
        j_end <- as.integer(end(ranges(mcols(neon12_annot_ers)[["grl"]])[a]))
        if (inbetween(value = j_start, rstart = er_start, rend = er_end) &
            inbetween(value = j_end, rstart = er_start, rend = er_end)) {
            er_start <- j_start + 1
            er_end <- j_end - 1
            change <- TRUE
        } else if (inbetween(value = j_start, rstart = er_start, rend = er_end)) {
            er_end <- j_start - 1
            change <- TRUE
        } else if (inbetween(value = j_end, rstart = er_start, rend = er_end)) {
            er_start <- j_end + 1
            change <- TRUE
        }
    } else if (length(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]]) == 2) {
        if (colgrs(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]])) {
            changes <- c(changes, change)
            next
        }
        if (ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][1] >
            ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][2]) {
            er_start <- as.integer(end(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][2])) + 1
            er_end <- as.integer(start(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][1])) - 1
            change <- TRUE
        } else {
            er_start <- as.integer(end(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][1])) + 1
            er_end <- as.integer(start(ranges(mcols(neon12_annot_ers)[["grl"]])[[a]][2])) - 1
            change <- TRUE
        }
    }

    start(check_neon12_annot_ers[a]) <- er_start
    end(check_neon12_annot_ers[a]) <- er_end
    changes <- c(changes, change)
}


# for (a in seq_along(ne_annot_ers)){ #takes around 2mins
#
#
#
#   er_start <-  start(ranges(ne_annot_ers)[a])
#   er_end <-  end(ranges(ne_annot_ers)[a])
#   if (length(ranges(mcols(ne_annot_ers)[["grl"]])[[a]])==1){
#     j_start <- start(ranges(mcols(ne_annot_ers)[["grl"]])[a])
#     j_end <- end(ranges(mcols(ne_annot_ers)[["grl"]])[a])
#   }
# }

# for (a in seq_along(neon_annot_ers)){ #running for 15mins and still not finished
#   er_start <-  start(neon_annot_ers[a])
#   er_end <-  end(neon_annot_ers[a])
#   if (length(mcols(neon_annot_ers)[["grl"]][[a]])==1){
#     j_start <- start(mcols(neon_annot_ers)[["grl"]][[a]])
#     j_end <- end(mcols(neon_annot_ers)[["grl"]][[a]])
#   }
# }
```

# Reproducibility


This package was developed using `r BiocStyle::Biocpkg("biocthis")`.


Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library("rmarkdown")
system.time(render("ODER.Rmd", "BiocStyle::html_document"))

## Extract the R code
library("knitr")
knit("ODER.Rmd", tangle = TRUE)
```

Date the vignette was generated.

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wallclock time spent generating the vignette.

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

`R` session information.

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```



# Bibliography


```{r vignetteBiblio, results = "asis", echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
# PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
